pkg/controller/uniteddeployment/uniteddeployment_update.go

```diff
@@ -0,0 +1,151 @@
+	expectedSubsets := sets.String{}
+	gotSubsets := sets.String{}
+
+	for _, subset := range ud.Spec.Topology.Subsets {
+		expectedSubsets.Insert(subset.Name)
+	}
+
+	for subsetName := range nameToSubset {
+		gotSubsets.Insert(subsetName)
+	}
+	klog.V(4).Infof("UnitedDeployment %s/%s has subsets %v, expects subsets %v", ud.Namespace, ud.Name, gotSubsets.List(), expectedSubsets.List())
+
+	var creates []string
+	for _, expectSubset := range expectedSubsets.List() {
+		if gotSubsets.Has(expectSubset) {
+			continue
+		}
+
+		creates = append(creates, expectSubset)
+	}
+
+	var deletes []string
+	for _, gotSubset := range gotSubsets.List() {
+		if expectedSubsets.Has(gotSubset) {
+			continue
+		}
+
+		deletes = append(deletes, gotSubset)
+	}
+
+	var errs []error
+	// manage creating
+	if len(creates) > 0 {
+		// do not consider deletion
+		klog.V(0).Infof("UnitedDeployment %s/%s needs creating subset (%s) with name: %v", ud.Namespace, ud.Name, subsetType, creates)
+		creatdChan := make(chan string, len(creates))
+		defer close(creatdChan)
+		for _, subset := range creates {
+			creatdChan <- subset
+		}
+
+		revision := currentRevision.Name
+		if updateRevision != nil {
+			revision = updateRevision.Name
+		}
+
+		var createdNum int
+		var createdErr error
+		createdNum, createdErr = util.SlowStartBatch(len(creates), slowStartInitialBatchSize, func() error {
+			subsetName := <-creatdChan
+
+			replicas := nextReplicas[subsetName]
+			partition := nextPartitions[subsetName]
+			err := r.subSetControls[subsetType].CreateSubset(ud, subsetName, revision, replicas, partition)
+			if err != nil {
+				if !errors.IsTimeout(err) {
+					return fmt.Errorf("fail to create Subset (%s) %s: %s", subsetType, subsetName, err.Error())
+				}
+			}
+
+			return nil
+		})
+		if createdErr == nil {
+			r.recorder.Eventf(ud.DeepCopy(), corev1.EventTypeNormal, fmt.Sprintf("Successful%s", eventTypeSubsetsUpdate), "Create %d Subset (%s)", createdNum, subsetType)
+		} else {
+			errs = append(errs, createdErr)
+		}
+	}
+
+	// manage deleting
+	if len(deletes) > 0 {
+		klog.V(0).Infof("UnitedDeployment %s/%s needs deleting subset (%s) with name: [%v]", ud.Namespace, ud.Name, subsetType, deletes)
+		var deleteErrs []error
+		for _, subsetName := range deletes {
+			subset := nameToSubset[subsetName]
+			if err := r.subSetControls[subsetType].DeleteSubset(subset); err != nil {
+				deleteErrs = append(deleteErrs, fmt.Errorf("fail to delete Subset (%s) %s/%s for %s: %s", subsetType, subset.Namespace, subset.Name, subsetName, err))
+			}
+		}
+
+		if len(deleteErrs) > 0 {
+			errs = append(errs, deleteErrs...)
+		} else {
+			r.recorder.Eventf(ud.DeepCopy(), corev1.EventTypeNormal, fmt.Sprintf("Successful%s", eventTypeSubsetsUpdate), "Delete %d Subset (%s)", len(deletes), subsetType)
+		}
+	}
+
+	// clean the other kind of subsets
```

Comment:
Why do we need this logic?

Reply:
To support changing subsetType.

If subset tyoe changed, ud has to delete the other kind of subset.